### PR TITLE
Solution for CF-718C

### PR DESCRIPTION
**Pull request by dutinmeow**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._
- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

**Comment by dutinmeow** [comment]
go away

File: solutions/cf-718C.mdx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+			res = multiply(res, base);
+		}
+		base = multiply(base, base);
+	}
+	return res;
+}
+
+int N, Q;
+pll tree[400005];
+ll lazy[400005];
```
Comment (Aadit-Ambadkar) [inline solutions/cf-718C.mdx:94]:
```suggestion
const int MAX = 400005;

pll tree[MAX];
ll lazy[MAX];
```

File: solutions/cf-718C.mdx [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	pushdown(t);
+
+	int tm = (tl + tr) >> 1;
+	return (query(l, r, t << 1, tl, tm) + query(l, r, t << 1 | 1, tm + 1, tr)) % MOD;
+}
+
+int main() {
+	cin >> N >> Q;
+	fill_n(tree, 400005, pll{0, 1});
```
Comment (Aadit-Ambadkar) [inline solutions/cf-718C.mdx:142]:
```suggestion
	fill_n(tree, MAX, pll{0, 1});
```

File: solutions/cf-718C.mdx (hunk 1 above)
Comment (dutinmeow) [reply solutions/cf-718C.mdx]: done

File: solutions/cf-718C.mdx [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+pll multiply(const pll &a, const Matrix &b) {
+	return {(a.FF * b[0][0] + a.SS * b[1][0]) % MOD, 
+		(a.FF * b[0][1] + a.SS * b[1][1]) % MOD};
+}
+
+Matrix pow(ll n) { 
+	Matrix res = {1, 0, 0, 1};
+	Matrix base = {0, 1, 1, 1};
+	for (; n; n >>= 1) {
```
Comment (Aadit-Ambadkar) [inline solutions/cf-718C.mdx:84]: explain this

**Review by Aadit-Ambadkar — APPROVED** [review]

File: solutions/cf-718C.mdx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+using namespace std;
+
+#define FF first
+#define SS second
+
+using ll = long long;
+using pll = pair<ll, ll>;
+using Matrix = array<array<ll, 2>, 2>;
+
+const int MAXN = 1e5 + 5;
```
Comment (SansPapyrus683) [inline solutions/cf-718C.mdx:66]: replace w/ 1e5 + 1
```suggestion
const int MAXN = 1e5 + 1;
```

File: solutions/cf-718C.mdx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define FF first
+#define SS second
```
Comment (SansPapyrus683) [inline solutions/cf-718C.mdx:60]: use only the crap here
https://github.com/cpinitiative/usaco-guide/issues/806

File: solutions/cf-718C.mdx [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+using ll = long long;
+using pll = pair<ll, ll>;
+using Matrix = array<array<ll, 2>, 2>;
+
+const int MAXN = 1e5 + 5;
+const ll MOD = 1e9 + 7;
+
+// multiplies two two-by-two matricies
```
Comment (SansPapyrus683) [inline solutions/cf-718C.mdx:69]: wrap matrix functions in a codesnip

File: solutions/cf-718C.mdx [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+pll multiply(const pll &a, const Matrix &b) {
+	return {(a.FF * b[0][0] + a.SS * b[1][0]) % MOD, 
+		(a.FF * b[0][1] + a.SS * b[1][1]) % MOD};
+}
+
+// returns {0, 1, 1, 1,} to the power of n
+Matrix pow(ll n) { 
+	Matrix res = {1, 0, 0, 1};  // identity matrix
+	Matrix base = {0, 1, 1, 1};  // fibonacci matrix
+	for (; n; n >>= 1) {. // binary exponentiation (log n)
```
Comment (SansPapyrus683) [inline solutions/cf-718C.mdx:87]: dustin did you mean to put a period here

File: solutions/cf-718C.mdx [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+	while (Q--) {
+		int t; 
+		cin >> t;
+		if (t == 1) {
+			int l, r; 
+			ll v;
+			cin >> l >> r >> v;
+			update(l, r, v);
+		}
+		else if (t == 2) {
```
Comment (SansPapyrus683) [inline solutions/cf-718C.mdx:166]:
```suggestion
		} else if (t == 2) {
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cf-718C.mdx (hunk 5 above)
Comment (dutinmeow) [reply solutions/cf-718C.mdx:60]: well good thing this isn't bronze through gold then

Comment (SansPapyrus683) [reply solutions/cf-718C.mdx:60]: I hate you so much.
image

\lh

File: solutions/cf-718C.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply solutions/cf-718C.mdx:90]: uuuhhh

**Review by SansPapyrus683 — APPROVED** [review]

**Comment by dutinmeow** [comment]
@SansPapyrus683 
can this be merged?

**Comment by SansPapyrus683** [comment]
get an admin to review editorial and they can merge

File: solutions/cf-718C.mdx [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+## Explanation
+
+Firstly, a note on notation: let $F_i$ denote the $i$th fibonacci number. That is,
+
+$$
+F_i = 
+\begin{cases} 
+	0 & i = 0 \\ 
+	1 & i = 1 \\
+	F_{i - 1} + F{i - 2} & i \geq 2 \\
```
Comment (bqi343) [inline solutions/cf-718C.mdx:17]:
```suggestion
	F_{i - 1} + F_{i - 2} & i \geq 2 \\
```

File: solutions/cf-718C.mdx [hunk 10]
```diff
@@ -0,0 +1,179 @@
+---
+id: cf-718C
+source: cf
```
Comment (bqi343) [inline solutions/cf-718C.mdx:3]:
```suggestion
source: CF
```

File: solutions/cf-718C.mdx [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+Inductively, we can cycle by $k$ by multiplying multiple times:
+
+$$
+\begin{pmatrix} F_{i - 1 + k} & F_{i + k} \end{pmatrix} = \begin{pmatrix} F_{i - 1} & F_i \end{pmatrix}\begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}^k
+$$
+
+We can use binary exponentiation to quickly compute powers of the two-by-two matrix. 
+
+Secondly, we need to utilize the distribute property of matrix multiplication to same-size matricies. That is, if $m_k$ is a one-by-two matrix for $k \in [1, n]$, then the following property is satisfied:
```
Comment (bqi343) [inline solutions/cf-718C.mdx:41]:
```suggestion
Secondly, we need to utilize the distributive property of matrix multiplication to same-size matricies. That is, if $m_k$ is a one-by-two matrix for $k \in [1, n]$, then the following property is satisfied:
```

Comment (bqi343) [reply solutions/cf-718C.mdx:41]:
```suggestion
Secondly, we need to utilize the distribute property of matrix multiplication to square matrices. That is, if $m_k$ is a one-by-two matrix for $k \in [1, n]$, then the following property is satisfied:
```

File: solutions/cf-718C.mdx [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+$$
+(m_1 + m_2 + \dots + m_n) \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} = m_1 \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} + m_2 \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} + \dots + m_n \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}
+$$
+
+In that way, we can utilize lazy propgation on the segment tree. We will store an integer tag in each node denoting the lazy update, which is denotes that we must cycle every node in its subtre by the value of the tag. When propogating, we can simply update the node by binary exponentiating and then multiplying, then storing it in the lazy tag. Querying works as normal on the segment tree.
```
Comment (bqi343) [inline solutions/cf-718C.mdx:47]:
```suggestion
In that way, we can utilize lazy propagation on the segment tree. We will store an integer tag in each node denoting the lazy update, which denotes that we must cycle every node in its subtree by the value of the tag. When propagating, we can simply update the node by binary exponentiating and then multiplying, then storing it in the lazy tag. Querying works as normal on the segment tree. 
```

File: solutions/cf-718C.mdx (hunk 7 above)
Comment (bqi343) [reply solutions/cf-718C.mdx:90]: ^^

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Comment by dutinmeow** [comment]
hopefully no spelling mistakes

**Review by bqi343 — APPROVED** [review]

**Pull request merged**